exercises/src/main/scala/fpinscala/datastructures/List.scala

```diff
@@ -46,23 +50,167 @@ object List { // `List` companion object. Contains functions for creating and wo
   def sum2(ns: List[Int]) = 
     foldRight(ns, 0)((x,y) => x + y)
   
+  /*
+  Product2, implemented in terms of foldRight, can't immediately halt
+  recursion on encountering a 0.0 because foldRight doesn't have any
+  logic for short-circuiting a recursive traversal of a list.
+
+  For short-circuiting to work, we would need to redefine foldRight to
+  pass it a 'sentinel' value; if it encounters the sentinel value in the
+  list, it would short-circuit and immediately return the accumulated
+  value.
+  */
   def product2(ns: List[Double]) = 
     foldRight(ns, 1.0)(_ * _) // `_ * _` is more concise notation for `(x,y) => x * y`; see sidebar
 
 
-  def tail[A](l: List[A]): List[A] = sys.error("todo")
+  def tail[A](l: List[A]): List[A] =
+    l match {
+      case Nil => ???
```

Comment:
sys.error("tail on empty list")

Sidenote: It's possible to avoid that:
https://github.com/scalaz/scalaz/blob/series/7.2.x/core/src/main/scala/scalaz/NonEmptyList.scala
https://github.com/idris-lang/Idris-dev/blob/master/libs/prelude/Prelude/List.idr#L87